File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int d = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        if(root==null){
            return 0;
        }

        int l = traverse(root.left,1);
        int r = traverse(root.right,1);

        return Math.max(d,l+r);
    }

    public int traverse(TreeNode root, int h){
        if(root==null){
            return 0;
        }

        int l = traverse(root.left,h+1);
        int r = traverse(root.right,h+1);
        
        d = Math.max(d,l+r);

        return Math.max(l,r)+1;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int cnt=1;
    int goodNodes(TreeNode* root) {
        traverse(root->left,root->val);
        traverse(root->right,root->val);

        return cnt;
    }
    void traverse(TreeNode* root,int h){
        if(root==NULL){
            return;
        }

        if(root->val>=h){
            cnt++;
            h=root->val;
        }

        traverse(root->left,h);
        traverse(root->right,h);
    }
};
